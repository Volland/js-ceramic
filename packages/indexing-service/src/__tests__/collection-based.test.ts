import 'reflect-metadata'

import { DataBase, IndexingService } from '../../lib'

function randomID(): string {
  return Math.random().toString(36).substring(2)
}

describe('collection-based queries', () => {
  const db = new DataBase({ type: 'sqlite', database: ':memory:', dropSchema: true })
  const service = new IndexingService({ db })

  afterAll(async () => {
    await db.dataSource.destroy()
  })

  test('get accounts using a definition in their index', async () => {
    // Need to be done in series to ensure snapshots are stable
    await service.index({
      id: 'recordID-1',
      tip: 'tip',
      metadata: {
        controllers: ['did:test:1'],
        family: 'definitionID-1',
        schema: 'ceramic://foo',
      },
    })
    await service.index({
      id: 'recordID-2',
      tip: 'tip',
      metadata: {
        controllers: ['did:test:2'],
        family: 'definitionID-1',
        schema: 'ceramic://foo',
      },
    })
    await service.index({
      id: 'recordID-3',
      tip: 'tip',
      metadata: {
        controllers: ['did:test:3'],
        family: 'definitionID-1',
        schema: 'ceramic://foo',
      },
    })

    // Shouldn't be in the results
    await Promise.all([
      service.index({
        id: randomID(),
        tip: 'tip',
        metadata: {
          controllers: ['did:test:1'],
          family: 'definitionID-1',
          // No schema
        },
      }),
      service.index({
        id: randomID(),
        tip: 'tip',
        metadata: {
          controllers: ['did:test:1'],
          // No family
          schema: 'ceramic://foo',
        },
      }),
      service.index({
        id: randomID(),
        tip: 'tip',
        metadata: {
          controllers: ['did:test:1'],
          family: 'definitionID-1',
          schema: 'ceramic://foo',
          unique: 'unique', // Non-deterministic
        },
      }),
      service.index({
        id: randomID(),
        tip: 'tip',
        metadata: {
          controllers: ['did:test:1'],
          family: 'definitionID-1',
          schema: 'ceramic://foo',
          tags: ['foo'], // Tags shouldn't be provided
        },
      }),
    ])

    await expect(
      service.getDefinitionAccountRecords({ definitionID: 'definitionID-1' })
    ).resolves.toMatchSnapshot()
  })
})
