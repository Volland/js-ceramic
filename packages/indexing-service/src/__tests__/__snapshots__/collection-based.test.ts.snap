// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`collection-based queries get accounts using a definition in their index 1`] = `
Object {
  "records": Array [
    Object {
      "account": "did:test:1",
      "recordID": "recordID-1",
    },
    Object {
      "account": "did:test:2",
      "recordID": "recordID-2",
    },
    Object {
      "account": "did:test:3",
      "recordID": "recordID-3",
    },
  ],
  "total": 3,
}
`;
